cibuildwheel/macos.py

```diff
@@ -242,60 +331,140 @@ def build(options: BuildOptions) -> None:
+                            log.warning(unwrap('''
+                                While arm64 wheels can be built on x86_64, they cannot be tested. The
+                                ability to test the arm64 wheels will be added in a future release of
+                                cibuildwheel, once Apple Silicon CI runners are widely available. To
+                                silence this warning, set `CIBW_TEST_SKIP: *-macosx_arm64`.
+                            '''))
+                        elif config.identifier.endswith('_universal2'):
+                            log.warning(unwrap('''
+                                While universal2 wheels can be built on x86_64, the arm64 part of them
+                                cannot currently be tested. The ability to test the arm64 part of a
+                                universal2 wheel will be added in a future release of cibuildwheel, once
+                                Apple Silicon CI runners are widely available. To silence this warning,
+                                set `CIBW_TEST_SKIP: *-macosx_universal2:arm64`.
+                            '''))
+                        else:
+                            raise RuntimeError('unreachable')
+
+                        # skip this test
+                        continue
+
+                    log.step('Testing wheel...' if testing_arch == machine_arch else f'Testing wheel on {testing_arch}...')
+
+                    # set up a virtual environment to install and test from, to make sure
+                    # there are no dependencies that were pulled in at build time.
+                    call(['pip', 'install', 'virtualenv', *dependency_constraint_flags], env=env)
+                    venv_dir = Path(tempfile.mkdtemp())
+
+                    arch_prefix = []
+                    if testing_arch != machine_arch:
+                        if machine_arch == 'arm64' and testing_arch == 'x86_64':
+                            # rosetta2 will provide the emulation with just the arch prefix.
+                            arch_prefix = ['arch', '-x86_64']
+                        else:
+                            raise RuntimeError("don't know how to emulate {testing_arch} on {machine_arch}")
+
+                    # define a custom 'call' function that adds the arch prefix each time
+                    def call_with_arch(args: Sequence[PathOrStr], **kwargs: Any) -> int:
+                        if isinstance(args, str):
+                            args = ' '.join(arch_prefix) + ' ' + args
+                        else:
+                            args = [*arch_prefix, *args]
+                        return call(args, **kwargs)
+
+                    # Use --no-download to ensure determinism by using seed libraries
+                    # built into virtualenv
+                    call_with_arch(['python', '-m', 'virtualenv', '--no-download', venv_dir], env=env)
```

Comment:
@joerick,

Have you been able to confirm that virtualenv and all the following calls indeed honor arch on M1 ?
It was not the case for i386 at some point in the past.

Reply:
Yes, the tests are checking that `platform.machine()` is correct inside TEST_COMMAND. They're not checked in CI yet, but I've run these locally on my M1 machine and they pass.